fbpic/particles/spin/spin_tracker.py

```diff
@@ -0,0 +1,276 @@
+        randomly generated to have the average values along all
+        axis specified by the user (see spin_distr below).
+
+        .. math::
+            \\frac{d\\boldsymbol{s}}{dt} = (\\boldsymbol{\\Omega}_T +
+             \\boldsymbol{\\Omega}_a) \\boldsymbol{s}
+
+        where
+
+        .. math::
+            \\boldsymbol{\\Omega}_T = \\frac{q}{m}\\left(
+                 \\frac{\\boldsymbol{B}}{\\gamma} -
+                 \\frac{\\boldsymbol{B}}{1+\\gamma}
+                 \\times \\frac{\\boldsymbol{E}}{c} \\right)
+
+        and
+
+        .. math::
+            \\boldsymbol{\\Omega}_a = a_e \\frac{q}{m}\\left(
+                 \\boldsymbol{B} -
+                 \\frac{\\gamma}{1+\\gamma}\\boldsymbol{\\beta}
+                 (\\boldsymbol{\\beta}\\cdot\\boldsymbol{B}) -
+                 \\boldsymbol{\\beta} \\times \\frac{\\boldsymbol{E}}{c} \\right)
+
+        Here, :math:`a_e` is the anomalous magnetic moment of the particle,
+        :math:`\\gamma` is the Lorentz factor of the particle,
+        :math:`\\boldsymbol{\\beta}=\\boldsymbol{v}/c` is the normalised velocity
+
+        The implementation of the push algorithm is detailed in
+        https://arxiv.org/abs/2303.16966.
+
+        Parameters
+        ----------
+        species: an fbpic Particles object
+            The species with which the spin tracker is
+            associated with
+
+        dt: float (in second)
+            Simulation timestep
+
+        sx_m: float (dimensionless), optional
+            The species-averaged average projection onto
+            the x-axis
+
+        sy_m: float (dimensionless), optional
+            The species-averaged average projection onto
+            the y-axis
+
+        sz_m: float (dimensionless), optional
+            The species-averaged average projection onto
+            the z-axis
+
+        anom: float, (dimensionless), optional
+            The anomalous magnetic moment of the particle,
+            given by :math:`a=(g-2)/2`, where :math:`g` is the
+            particle's g-factor.
+
+        spin_distr: str, optional
+            If 'fixed', all particles will have a fixed spin value
+            equal to s{x,y,z}_m.
+            If 'rand', the spin vectors will be random, but with an
+            ensemble average defined by one of the values of
+            s{x,y,z}_m. The first non-zero mean component will be
+            used, with order of preference being x,y,z, ie if sx_m!=0,
+            the generated spins will have an ensemble averages of
+            |sx|=sx_m, |sy|=0, |sz|=0, or if sz_m!=0, |sx|=0, |sy|=0
+            and |sz|=sz_m.
+
+        """
+        self.sx_m = sx_m
+        self.sy_m = sy_m
+        self.sz_m = sz_m
+        self.sm = np.sqrt(sx_m ** 2 + sy_m ** 2 + sz_m ** 2)
+        self.anom = anom
+        self.dt = dt
+        self.spin_distr = spin_distr
+
+        # Store the species we perform spin tracking for
+        self.ptcl = species
```

Comment:
Could this be avoided, instead passing `species` explicitly as an argument to `store_previous_momenta` and `push_s`?
(This is mostly for stylistic preferences ; but if it is too much of a hassle, it's also fine to leave this as is.)

Reply:
Good idea! It was a lot easier than I thought as well :)